gnomad/utils/constraint.py

```diff
@@ -360,3 +412,285 @@ def collapse_strand(
+    observed_variants_expr: hl.expr.Int64Expression,
+    possible_variants_expr: hl.expr.Int64Expression,
+    pop_observed_variants_array_expr: hl.expr.ArrayExpression = None,
+    weighted: bool = False,
+) -> Dict[str, Union[Dict[bool, hl.expr.ArrayExpression], hl.ArrayExpression]]:
+    """
+    Build plateau models to calibrate mutation rate to compute predicted proportion observed value.
+
+    The x and y of the plateau models:
+    - x: `mu_snp` - mutation rate
+    - y: proportion observed ('observed_variants' or 'observed_{pop}' / 'possible_variants')
+
+    :param cpg_expr: BooleanExpression noting whether a site is a CPG site.
+    :param mu_snp_expr: Float64Expression of the mutation rate.
+    :param observed_variants_expr: Int64Expression of the observed variant counts
+        for each combination of keys in `ht`.
+    :param possible_variants_expr: Int64Expression of the possible variant counts
+        for each combination of keys in `ht`.
+    :param pop_observed_variants_array_expr: Nested ArrayExpression with all observed
+        variant counts ArrayNumericExpressions for specified populations. e.g., `[[1,1,
+        1],[1,1,1]]`. Default is None.
+    :param weighted: Whether to generalize the model to weighted least squares using
+        'possible_variants'. Default is False.
+    :return: A dictionary of intercepts and slopes for plateau models of each
+    population. The keys for this dictionary are 'total' and 'pop'. The values for
+    'total' is a dictionary (e.g., <DictExpression of type dict<bool,
+    array<float64>>>), and the value for 'pop' is a nested list of dictionaries (e.g.,
+    <ArrayExpression of type array<array<dict<bool, array<float64>>>>>). The key of the
+    dictionary in the nested list is CpG status (BooleanExpression), and the value is
+    an ArrayExpression containing intercept and slope values.
+    """
+    # Build a plateau model using all the sites in the Table.
+    plateau_models_agg_expr = {
+        "total": hl.agg.group_by(
+            cpg_expr,
+            hl.agg.linreg(
+                observed_variants_expr / possible_variants_expr,
+                [1, mu_snp_expr],
+                weight=possible_variants_expr if weighted else None,
+            ).beta,
+        )
+    }
+    if pop_observed_variants_array_expr is not None:
+        # Build plateau models using sites in population downsamplings if
+        # population is specified.
+        plateau_models_agg_expr["pop"] = hl.agg.array_agg(
+            lambda pop_observed_variants_expr: hl.agg.array_agg(
+                lambda pop_observed_variants: hl.agg.group_by(
+                    cpg_expr,
+                    hl.agg.linreg(
+                        pop_observed_variants / possible_variants_expr,
+                        [1, mu_snp_expr],
+                        weight=possible_variants_expr,
+                    ).beta,
+                ),
+                pop_observed_variants_expr,
```

Comment:
technically this is also an array. I wonder if there are more descriptive variable names you could use in this agg?